### ISSUE TITLE
Changes in behaviour due to buffer protocol in numpy 1.5 (migrated from Trac #1312)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2865]
Original ticket http://projects.scipy.org/numpy/ticket/1312
Reported 2009-12-06 by atmention:GaelVaroquaux, assigned to atmention:pv.

The following used to work in 1.4:
    hashlib.md5(x).hexdigest()
Now, you have to do:
    hashlib.md5(np.getbuffer(x)).hexdigest()

Also, with extension code I get the following failure (if not rebuild against a new version of numpy):
    from routines import _joint_histogram, _similarity, similarity_measures
      File "numpy.pxi", line 74, in nipy.neurospin.register.routines (nipy/neurospin/register/routines.c:6042)
    ValueError: numpy.dtype does not appear to be the correct type object
Where routines.c is a cython-generated file, compiled with numpy 1.4. I am unsure which line in routines.pyx is responsible for the problem.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-12-06

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-12-06

Something's strange in the state of python-2.6/Python/getargs.c:1318
    static Py_ssize_t
    convertbuffer(PyObject _arg, void *_p, char **errmsg)
    {
            PyBufferProcs *pb = arg->ob_type->tp_as_buffer;
            Py_ssize_t count;
            if (pb == NULL ||
                pb->bf_getreadbuffer == NULL ||
                pb->bf_getsegcount == NULL ||
                pb->bf_releasebuffer != NULL) {
                    *errmsg = "string or read-only buffer";
                    return -1;
            }
Apparently, PyArg_ParseTuple, when handling "s#" that hashlib uses, cannot handle any buffers that (in the new buffer protocol) require releasing. The correct fix is probably to change hashlib to use "s#".

Some of the PyBuffer_\* methods actually work so that they release the buffer immediately after acquiring a char\* pointer. Unsafe, though...

The work-around is likely to not use `bf_releasebuffer` at all in ndarray (there are also other issues associated with it), and free any extra memory required by the buffer interface when the array is deallocated.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-12-06

Correction: "The correct fix is probably to change hashlib to use "s*"."

**Comment by thouis** [comment]
Comment in Trac by atmention:GaelVaroquaux, 2009-12-06

I more reproducible error, for the second problem: if scipy is not recompiled, I have:
    /home/varoquau/dev/scipy/scipy/stats/distributions.py in <module>()
         25 from scipy.special import gammaln as gamln
         26 from copy import copy
    ---> 27 import vonmises_cython
         28 import textwrap
         29 

```
/home/varoquau/Desktop/graph_francois/numpy.pxd in scipy.stats.vonmises_cython (scipy/stats/vonmises_cython.c:2939)()
     28 
     29 
---> 30 
     31 
     32 

ValueError: numpy.dtype does not appear to be the correct type object
```

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-12-06

The first issue is worked around in r7974; we just have to do without `bf_releasebuffer`. Please test if you have time.

The second issue is resolved by recompiling. There's really no way around this -- if we add new stuff to type objects, we can't maintain a fully backward compatible ABI. (Or actually, for the most part we can, Cython is just being very strict here.) I doubt there's a way around this, unless we drop PEP 3118 support.

The first issue needs still some thinking, though: mutability of dtypes makes the solution in r7974 not complete, so we'll keep this ticket open for now...

**Comment by thouis** [comment]
Comment in Trac by atmention:GaelVaroquaux, 2009-12-08

Replying to [comment:5 pv]:

> The first issue is worked around in r7974; we just have to do without `bf_releasebuffer`. Please test if you have time.

First issue is indeed fixed.

> The second issue is resolved by recompiling. There's really no way around this -- if we add new stuff to type objects, we can't maintain a fully backward compatible ABI. (Or actually, for the most part we can, Cython is just being very strict here.) I doubt there's a way around this, unless we drop PEP 3118 support.

Robert Bradshaw has contacted me for more information on the usecase and how Cython could be improved. Unfortunately, I am not the right person to talk to about this, because I haven't followed the changes in the underlying C code. I'll Cc you in the discussion.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-03-18

The first issue with "s#" was worked around in SVN trunk. Please reopen if something still breaks.

The Cython issue is addressed in new Cython versions, and Numpy 2.0 will anyway introduce an ABI break.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-06
